README.md
# NSCAPI: a server that ingests nsca calls and exposes the result via a json API

[![TravisBuild](https://travis-ci.org/tubemogul/nscapi.svg?branch=master)](https://travis-ci.org/tubemogul/nscapi)

## Introduction

Nscapi is a go application meant to help moving from a nsca model to a pull
model by providing a JSON endpoint to summarize the state of the nsca calls it
received. 

The tool is split into 3 main parts:

* Server that receive nsca calls and put them in a non-locking queue
* Worker that reads from the non-locking queue and put the data in a cache
* HTTP server that will read from the cache and display the check results 
  in a json format in response to a call on `/api/reports` based on a template.
  The rest of the calls will display a usage summary based on another template

## Usage

### Command-line options

To try to stick to the 12 factors principle, our configuration can be provider
from command-line arguments but also from environment variables. Note that the
command-line arguments have a higher priority than the environment variables.

```
Usage of ./nscapi:
  -api-ip string
    	IP the API should listen on. Default to the NSCAPI_API_IP environment variable. Fallback: 0.0.0.0 (default "0.0.0.0")
  -api-port uint
    	Port the API should listen on. Default to the NSCAPI_API_PORT environment variable. Fallback: 8080 (default 8080)

  -api-templates-root string
    	Root directory the API should use to look for its templates (root.tmpl and reports_element.tmpl). Default to the NSCAPI_API_TEMPLATES_ROOT environment variable. Fallback: templates (default "templates")

  -api-custom-fields-root string
    	Root directory the API should use as root of the custom fields hierarchy. Default to the NSCAPI_API_CUSTOM_FIELDS_ROOT environment variable. Fallback: custom_fields (default "custom_fields")

  -nsca-server-ip string
    	IP the NSCA server should listen on. Default to the NSCAPI_NSCA_IP environment variable. Fallback: 0.0.0.0 (default "0.0.0.0")
  -nsca-server-port uint
    	Port the NSCA server should listen on. Default to the NSCAPI_NSCA_PORT environment variable. Fallback: 5667 (default 5667)

  -nsca-server-password string
    	Password the NSCA server should use. Default to the NSCAPI_NSCA_PASSWORD environment variable. Fallback: ''

  -nsca-server-encryption uint
    	Number corresponding to the encryption to be used by the NSCA server. Default to the NSCAPI_NSCA_ENCRYPTION environment variable. Fallback: 0. See 'DECRYPTION METHOD' on https://github.com/NagiosEnterprises/nsca/blob/master/sample-config/nsca.cfg.in for more details. Must be <27.

```
The list of encryption algorithm code number can be found [here](https://github.com/NagiosEnterprises/nsca/blob/master/sample-config/nsca.cfg.in)

In this example, the API server will run on port 8081 (listening on all
available network interfaces), the NSCA server will be listening on localhost
on port 5668, without encryption:
```
export NSCAPI_NSCA_IP=localhost
./nscapi -nsca-server-port 5668
```
Note that we recommend not to mix the environment variables and command-line
arguments as it might end up being confusing. Just use either of them and stick
to it for readability (the application has no problem handling both but whoever
take back the work after might be confused).

## Custom fields

A custom field is a key-value couple that is not contained in the nsca check
itself but will be returned by the API based on the values of the check itself.

The custom fields are defined in a hiera-style hierarchy.

Provided that:
* `%{hostgroup}` is the hostname received in the nsca check result minus any trailing number
* `%{check name}` is the name of the check as received in the nsca check result

The level of importance is defined as follows:
* `common.yaml` is where the default values are defined.
* `service/%{hostgroup}/all.yaml` is where you define the default values of your
  custom fields per host group. It has the priority over the custom fields
  already defined in `common.yaml` and can also add new fields of its own.
  If a host don't match any of the defined hostgroups, only the custom fields
  defined in `common.yaml` will be applied.
* `service/%{hostgroup}/%{check name}.yaml` has the highest priority. Any field value in
  this file will overwrite the values of the previously defined custom fields.

## Using the Makefile

The Makefile is used as a helper for building and testing the project. Current
capabilities:

 * `make lint`: runs golint on your files (requires `github.com/golang/lint/golint` installed)
 * `make fmt`: checks that the files are compliant with the gofmt format
 * `make vet`: runs `go tool vet` on your files to ensure there's no problems
 * `make test`: runs `make lint`, `make fmt`, `make vet` before running all the
   test, printing also the percentage of code coverage
 * `make race`: runs the tests with the `-race` option to detect race conditions
 * `make bench`: runs the benchmarks
 * `make gocov`: runs a gocov report (requires `github.com/axw/gocov/gocov`)
 * `make dep`: gets (with update option `-u`) the go dependencies required by
   the application
 * `make build`: runs `make dep` and `make test` before running a clean and
   build the binary
 * `make install`: runs `make dep` and `make test` before running a clean and
   install the application
 * `make` / `make all`: run `make test` and `make race`

## Contributions

Contributions to this project are welcome, though please
[file an issue](https://github.com/tubemogul/nscapi/issues/new).
before starting work on anything major as someone else could already be working
on it.

Contributions that do not provide the corresponding tests will most likely not
be accepted (we can help building the tests if needed).

Contributions that do not pass the basic gofmt, vet and other basic checks
provided in the Makefile will not be accepted. It's just a question of trying to
keep a basic code standard. Thanks for your help! :)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2017-02-01
### Added
- Server that receive nsca calls and put them in a non-locking queue
- Worker that reads from the non-locking queue and put the data in a cache
- HTTP server that will read from the cache and display the check results
  in a json format in response to a call on `/api/reports` based on a template.
  The rest of the calls will display a usage summary based on another template
- Support for custom fields
- Support configuration via both command-line and environment variables


api_test.go
package main

import (
	"testing"
)

func TestStatusString(t *testing.T) {
	cases := []struct {
		in  int16
		out string
	}{{0, "OK"}, {1, "Warning"}, {2, "Critical"}, {3, "Unknown"}}
	for _, tt := range cases {
		returned := statusString(tt.in)
		if returned != tt.out {
			t.Errorf("statusString(%d) should return %s, not %s", tt.in, tt.out, returned)
		}
	}
}


main_test.go
package main

import (
	"github.com/PurpureGecko/go-lfc"
	nsca "github.com/tubemogul/nscatools"
	"testing"
)

func TestQueueData(t *testing.T) {
	q = lfc.NewQueue()
	if q.Len() > 0 {
		t.Error("Queue not empty on start")
	}

	p := &nsca.DataPacket{HostName: "host01", Service: "service foo", PluginOutput: "OK", Timestamp: 1484527962, State: 0}
	if err := queueData(p); err != nil {
		t.Errorf("queueData returned: %s", err)
	}
	if l := q.Len(); l != 1 {
		t.Errorf("Queue expecting to contain 1 element. Contains %d", l)
	}
}

func TestCacheWorker(t *testing.T) {
	initCache()
	q = lfc.NewQueue()
	if q.Len() > 0 {
		t.Error("Queue not empty on start")
	}

	testCases := []struct {
		host      string
		service   string
		output    string
		timestamp uint32
		state     int16
	}{
		{"host01", "service foo", "Bar", 1484527962, 0},
		// Update the same service entry as created previously
		{"host01", "service foo", "New output", 1484527963, 1},
		// Second service on the same host
		{"host01", "service bar", "OK", 1484527964, 0},
		// Another host
		{"host02", "service whatever", "OK", 1484527965, 0},
		// 2nd service with the same name as the other host
		{"host02", "service bar", "OK", 1484527966, 0},
	}

	for _, tt := range testCases {
		p := &nsca.DataPacket{HostName: tt.host, Service: tt.service, PluginOutput: tt.output, Timestamp: tt.timestamp, State: tt.state}
		q.Enqueue(p)
		cacheWorker(false)
		h, ok := cache[tt.host]
		if !ok {
			t.Errorf("Entry %s should be present", tt.host)
		}
		s, ok := h[tt.service]
		if !ok {
			t.Errorf("%s should still have a %s entry", tt.host, tt.service)
		}
		if s.timestamp != tt.timestamp {
			t.Errorf("entry '%s' has wrong timestamp. Got %d, expecting %d", tt.service, s.timestamp, tt.timestamp)
		}
		if s.output != tt.output {
			t.Errorf("entry '%s' has wrong output. Got %s, expecting %s", tt.service, s.output, tt.output)
		}
		if s.state != tt.state {
			t.Errorf("entry '%s' has wrong state. Got %d, expecting %d", tt.service, s.state, tt.state)
		}
	}
}


cache_test.go
package main

import "testing"

func TestInitCache(t *testing.T) {
	if cache != nil {
		t.Errorf("Already initialized cache found")
	}
	initCache()
	if cache == nil {
		t.Errorf("Cache object still not initialized after the init")
	}
}

func TestUpdateCacheEntry(t *testing.T) {
	initCache()

	// Create new entry
	h, ok := cache["host01"]
	if ok {
		t.Errorf("Entry host01 already exists")
	}
	updateCacheEntry("host01", "service foo", "OK", 1484527962, 0)
	h, ok = cache["host01"]
	if !ok {
		t.Errorf("Entry host01 should have been created")
	}
	s, ok := h["service foo"]
	if !ok {
		t.Errorf("host01 should have a service foo entry")
	}
	if s.timestamp != 1484527962 {
		t.Errorf("service foo has wrong timestamp upon creation")
	}
	if s.output != "OK" {
		t.Errorf("service foo has wrong output upon creation")
	}
	if s.state != 0 {
		t.Errorf("service foo has wrong state upon creation")
	}

	testData := []struct {
		host      string
		service   string
		output    string
		timestamp uint32
		state     int16
	}{
		// Update the same service entry as created previously
		{"host01", "service foo", "New output", 1484527963, 1},
		// Second service on the same host
		{"host01", "service bar", "OK", 1484527964, 0},
		// Another host
		{"host02", "service whatever", "OK", 1484527965, 0},
		// 2nd service with the same name as the other host
		{"host02", "service bar", "OK", 1484527966, 0},
	}
	// Launch all the updates before running the tests
	for _, tt := range testData {
		updateCacheEntry(tt.host, tt.service, tt.output, tt.timestamp, tt.state)
	}

	for _, tt := range testData {
		h, ok = cache[tt.host]
		if !ok {
			t.Errorf("Entry %s should be present", tt.host)
		}
		s, ok = h[tt.service]
		if !ok {
			t.Errorf("%s should still have a %s entry", tt.host, tt.service)
		}
		if s.timestamp != tt.timestamp {
			t.Errorf("entry '%s' has wrong timestamp. Got %d, expecting %d", tt.service, s.timestamp, tt.timestamp)
		}
		if s.output != tt.output {
			t.Errorf("entry '%s' has wrong output. Got %s, expecting %s", tt.service, s.output, tt.output)
		}
		if s.state != tt.state {
			t.Errorf("entry '%s' has wrong state. Got %d, expecting %d", tt.service, s.state, tt.state)
		}
	}
}


api.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"text/template"
)

var tmplRoot string

// statusString returns the corresponding string to the nagios status
func statusString(state int16) string {
	switch state {
	case 0:
		return "OK"
	case 1:
		return "Warning"
	case 2:
		return "Critical"
	}
	return "Unknown"
}

// ToJSONString is a function that takes an interface as output and formats it for
// json output in return. Its purpose is to be used in the templates which is
// the reason wly it is exported
func ToJSONString(v interface{}) string {
	bytesOutput, _ := json.Marshal(v)
	return string(bytesOutput)
}

// rootHandler just renders the root.tmpl that explains the api calls usage
func rootHandler(w http.ResponseWriter, r *http.Request) {
	tmplPath := filepath.Join(tmplRoot, "root.tmpl")
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	t := template.Must(template.ParseFiles(tmplPath))
	t.Execute(w, nil)
}

// reportsHandler takes care of the path /api/reports that lists all the checks
// on all the hosts, each elements defined based on the reports_element.tmpl
// template
func reportsHandler(w http.ResponseWriter, r *http.Request) {
	tmplName := "reports_element.tmpl"
	tmplPath := filepath.Join(tmplRoot, tmplName)
	w.Header().Set("Content-Type", "application/json")
	fMaps := template.FuncMap{"tojson": ToJSONString}
	t := template.Must(template.New(tmplName).Funcs(fMaps).ParseFiles(tmplPath))
	io.WriteString(w, "[")
	i := 1
	for host, svcs := range cache {
		j := 1
		for svc, chk := range svcs {
			c := map[string]map[string]interface{}{
				"check": map[string]interface{}{"host": host, "name": svc, "status": statusString(chk.state), "message": chk.output, "timestamp": fmt.Sprint(chk.timestamp)},
				// custom will be used to inject custom-defined fields
				"custom": getCustomFields(host, svc),
			}
			t.Execute(w, c)
			// This part just takes care of adding a coma or not between the elements
			// to have a correcly-formated json
			if !(i == len(cache) && j == len(svcs)) {
				io.WriteString(w, ",")
			}
			j++
		}
		i++
	}
	io.WriteString(w, "]\n")
}

// setIfPathExists validates that the given path exists before assigning it to
// the given variable
func setIfPathExists(dir string, varToSet *string) error {
	if _, err := os.Stat(dir); err != nil {
		return err
	}
	*varToSet = dir
	return nil
}

// initAPIServer starts the API HTTP server. This is where the routes are
// defined. The customFieldRoot is the root of the hierarchy of yaml files used
// for the custom fields. The templatesRoot is the root directory where to find
// the templates used by the API
func initAPIServer(listenerIP string, port uint, customFieldRoot string, templatesRoot string) {
	// Init custom fields
	var customFRoot string
	setIfPathExists(customFieldRoot, &customFRoot)
	loadCustomFields(customFRoot)

	setIfPathExists(templatesRoot, &tmplRoot)
	http.HandleFunc("/", rootHandler)
	http.HandleFunc("/api/reports", reportsHandler)
	http.ListenAndServe(fmt.Sprint(listenerIP, ":", port), nil)
}


cache.go
package main

// The cache structure contains 2 layers of maps:
// * Layer 1: the key is the hostname
// * Layer 2: for each hostname, there's a map where the key is the service name
var cache map[string]map[string]*serviceEntry

// ServiceEntry can be found in the 2nd layer of he map and contains the details
// of the last status of the check (timestamp, state and plugin output)
type serviceEntry struct {
	timestamp uint32
	state     int16
	output    string
}

// initCache initialize the cache object
func initCache() {
	cache = make(map[string]map[string]*serviceEntry)
}

// updateCacheEntry adds or update a given service check result in the cache map
func updateCacheEntry(hostname, servicename, output string, timestamp uint32, state int16) {
	svc, ok := cache[hostname]
	if !ok {
		svc = make(map[string]*serviceEntry)
		cache[hostname] = svc
	}
	svc[servicename] = &serviceEntry{
		timestamp: timestamp,
		output:    output,
		state:     state,
	}
}


custom.go
package main

import (
	"gopkg.in/yaml.v2"
	"io/ioutil"
	"os"
	"path/filepath"
	"regexp"
)

// customFields is a map containing the entire custom fields hierarchy using the
// following format: customFields[fieldClassifier{hostgroup: hostgroup, service: checkName}][fieldName]=value.
// checkName can be "all" for the defaults at hostgroup and common level.
// The content of the common.yaml will use the following format
// customFields[fieldClassifier{hostgroup: "##common##", service: "all"}][fieldName]=value
var customFields map[fieldClassifier]map[string]interface{}

// fieldClassifier is used as the key in the customFields
type fieldClassifier struct {
	hostgroup, service string
}

// loadCustomFields loads in memory the custom fields based on the yaml
// hierarchy on disk
func loadCustomFields(rootPath string) error {
	customFields = make(map[fieldClassifier]map[string]interface{})
	if _, err := os.Stat(rootPath); err != nil {
		return err
	}
	// 1st common.yaml. Just ignore the errors for now
	commonFields, _ := processYamlFile(filepath.Join(rootPath, "common.yaml"))
	customFields[fieldClassifier{"##common##", "all"}] = make(map[string]interface{})
	if commonFields != nil {
		customFields[fieldClassifier{"##common##", "all"}] = commonFields
	}
	// Then service/hostgroup/*.yaml. Just ignore the errors for now
	files, _ := filepath.Glob(filepath.Join(rootPath, "service", "*", "*.yaml"))
	for _, f := range files {
		key := fieldClassifier{filepath.Base(filepath.Dir(f)), filepath.Base(f[:len(f)-5])}
		fields, _ := processYamlFile(f)
		customFields[key] = make(map[string]interface{})
		if fields != nil {
			customFields[key] = fields
		}
	}
	return nil
}

// processYamlFile tries to process a yaml file return its content hashed and/or
// an error
func processYamlFile(path string) (map[string]interface{}, error) {
	returnValue := make(map[string]interface{})
	fc, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, err
	}
	err = yaml.Unmarshal(fc, &returnValue)

	return returnValue, err
}

// mergeMapFromCustomFields merges the data from a given customFields key (aka
// yaml file loaded from the disk) into the map given as reference
func mergeMapFromCustomFields(reference map[string]interface{}, key *fieldClassifier) {
	if m, ok := customFields[*key]; ok {
		for f := range m {
			reference[f] = m[f]
		}
	}
}

// getCustomFields returns a hash containing the custom fields specific for this
// hostname and checkName
func getCustomFields(hostname, checkName string) map[string]interface{} {
	re, _ := regexp.Compile("[^A-Za-z0-9]$")
	hostgroup := re.ReplaceAllString(hostname, "")
	resultFields := make(map[string]interface{})
	mergeMapFromCustomFields(resultFields, &fieldClassifier{"##common##", "all"})
	mergeMapFromCustomFields(resultFields, &fieldClassifier{hostgroup, "all"})
	mergeMapFromCustomFields(resultFields, &fieldClassifier{hostgroup, checkName})
	return resultFields
}


main.go
package main

import (
	"flag"
	"github.com/PurpureGecko/go-lfc"
	nsca "github.com/tubemogul/nscatools"
	"os"
	"strconv"
	"time"
)

var q *lfc.Queue

type cfg struct {
	apiIP              string
	apiPort            uint
	apiCustomFieldRoot string
	apiTemplatesRoot   string
	nscaIP             string
	nscaPort           uint
	nscaPassword       string
	nscaEncryption     uint
}

// cacheWorker will continuously pull DataPackets out of the queue and update
// the cache with it. Passing true as parameter will make it loop indefinetly.
// Passing false as parameter will make it return whenever the queue is empty
func cacheWorker(runIndefinetly bool) {
	for {
		if q.Len() == 0 {
			if runIndefinetly {
				// Wait 100ms before checking for data in the queue
				time.Sleep(100 * time.Millisecond)
			} else {
				return
			}
		} else {
			pkt, ok := q.Dequeue()
			if ok {
				if p, ok := pkt.(*nsca.DataPacket); ok {
					updateCacheEntry(p.HostName, p.Service, p.PluginOutput, p.Timestamp, p.State)
				}
			}
		}
	}
}

// queueData will put the DataPacket received by the nsca server in a
// non-locking queue
func queueData(p *nsca.DataPacket) error {
	q.Enqueue(p)
	return nil
}

// getStringFromEnv gets the string value of the specified environment variable
// or the default value if this variable is not set
func getStringFromEnv(varName string, defaultValue string) string {
	var (
		val     string
		present bool
	)
	if val, present = os.LookupEnv(varName); !present {
		val = defaultValue
	}
	return val
}

// getUintFromEnv gets the uint value of the specified environment variable
// or the default value if this variable is not set
func getUintFromEnv(varName string, defaultValue uint, bitSize int) uint {
	var (
		strVal  string
		val     uint
		present bool
	)
	strVal, present = os.LookupEnv(varName)
	if val64, err := strconv.ParseUint(strVal, 10, bitSize); !present || err != nil {
		val = defaultValue
	} else {
		val = uint(val64)
	}
	return val
}

// initConfig initializes the configuration based on environment variables and
// passed flags (in the 12 factor app spirit)
func initConfig() *cfg {
	var conf cfg
	flag.StringVar(&conf.apiIP, "api-ip", getStringFromEnv("NSCAPI_API_IP", "0.0.0.0"), "IP the API should listen on. Default to the NSCAPI_API_IP environment variable. Fallback: 0.0.0.0")
	flag.UintVar(&conf.apiPort, "api-port", getUintFromEnv("NSCAPI_API_PORT", 8080, 32), "Port the API should listen on. Default to the NSCAPI_API_PORT environment variable. Fallback: 8080")
	flag.StringVar(&conf.apiCustomFieldRoot, "api-custom-fields-root", getStringFromEnv("NSCAPI_API_CUSTOM_FIELDS_ROOT", "custom_fields"), "Root directory the API should use as root of the custom fields hierarchy. Default to the NSCAPI_API_CUSTOM_FIELDS_ROOT environment variable. Fallback: custom_fields")
	flag.StringVar(&conf.apiTemplatesRoot, "api-templates-root", getStringFromEnv("NSCAPI_API_TEMPLATES_ROOT", "templates"), "Root directory the API should use to look for its templates (root.tmpl and reports_element.tmpl). Default to the NSCAPI_API_TEMPLATES_ROOT environment variable. Fallback: templates")
	flag.StringVar(&conf.nscaIP, "nsca-server-ip", getStringFromEnv("NSCAPI_NSCA_IP", "0.0.0.0"), "IP the NSCA server should listen on. Default to the NSCAPI_NSCA_IP environment variable. Fallback: 0.0.0.0")
	flag.UintVar(&conf.nscaPort, "nsca-server-port", getUintFromEnv("NSCAPI_NSCA_PORT", 5667, 16), "Port the NSCA server should listen on. Default to the NSCAPI_NSCA_PORT environment variable. Fallback: 5667")
	flag.StringVar(&conf.nscaPassword, "nsca-server-password", getStringFromEnv("NSCAPI_NSCA_PASSWORD", ""), "Password the NSCA server should use. Default to the NSCAPI_NSCA_PASSWORD environment variable. Fallback: ''")
	flag.UintVar(&conf.nscaEncryption, "nsca-server-encryption", getUintFromEnv("NSCAPI_NSCA_ENCYPTION", 0, 8), "Number corresponding to the encryption to be used by the NSCA server. Default to the NSCAPI_NSCA_ENCRYPTION environment variable. Fallback: 0. See 'DECRYPTION METHOD' on https://github.com/NagiosEnterprises/nsca/blob/master/sample-config/nsca.cfg.in for more details. Must be <27.")
	flag.Parse()
	return &conf
}

func main() {
	q = lfc.NewQueue()
	initCache()

	// Loads config from flags or from env
	srvConf := initConfig()

	// Start the worker that updates the cache
	go cacheWorker(true)

	// Start the API inside a routine
	go initAPIServer(srvConf.apiIP, srvConf.apiPort, srvConf.apiCustomFieldRoot, srvConf.apiTemplatesRoot)

	// Start the nsca server
	nscaCfg := nsca.NewConfig(srvConf.nscaIP, uint16(srvConf.nscaPort), int(srvConf.nscaEncryption), srvConf.nscaPassword, queueData)
	nsca.StartServer(nscaCfg, false)

}
